apidocs.yaml

```diff
@@ -0,0 +1,826 @@
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/accept-terms:
+    post:
+      summary: Accept terms of use
+      operationId: accept_application_terms_v1
+      tags:
+        - "terms"
+      parameters:
+        - name: id
+          in: path
+          description: ID of application to accept terms of use for
+          required: true
+          schema:
+            type: string
+      responses:
+        "204":
+          description: Successful Response (no content)
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/add-attachment:
+    post:
+      summary: Add attachment to application
+      operationId: add_attachment_to_application_v1
+      tags:
+        - "attachments"
+      parameters:
+        - name: id
+          in: path
+          description: ID of application to attach file to
+          required: true
+          schema:
+            type: string
+      requestBody:
+        content:
+          application/pdf:
+            schema:
+              type: string
+              format: binary
+      responses:
+        "200":
+          description: Successful Response
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: "#/components/schemas/AddedAttachments"
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/attachments/{attachmentId}:
+    get:
+      summary: Retrieve attachment from application
+      operationId: retrieve_attachment_from_application_v1
+      tags:
+        - "attachments"
+      parameters:
+        - name: id
+          in: path
+          description: ID of application to retrieve attachment from
+          required: true
+          schema:
+            type: string
+        - name: attachmentId
+          in: path
+          description: ID of attachment to retrieve
+          required: true
+          schema:
+            type: string
+      responses:
+        "200":
+          description: Successful Response
+          content:
+            application/pdf:
+              schema:
+                type: string
+                format: binary
+  /api/v1/applications/{id}/remove-attachment:
+    post:
+      summary: Remove attachment from application
+      operationId: remove_attachment_from_application_v1
+      tags:
+        - "attachments"
+      parameters:
+        - name: id
+          in: path
+          description: ID of application to remove attachment from
+          required: true
+          schema:
+            type: string
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/DeleteAttachment"
+      responses:
+        "204":
+          description: Successful Response (no content)
+      security:
+        - auth:
+          - write:applications
+  /api/v1/applications/{id}/add-member:
```

Comment:
POST `/api/v1/applications/{id}/members`

Then you can also have GET `/api/v1/applications/{id}/members` to get all members for application no?

Reply:
I must discuss this face to face with the team, but the reasons are multiple.

The problem: A RESTfull API is a basic CRUD, but it is too rigid and aggregates a lot of different responsibilities into too few endpoints, which very often do not reflect business scenarios and use cases. It's almost like a one to one between HTTP call and database queries.

So I wanted to prepare the ground and introduce different principles: event-driven design, DDD, Clean Archictecture, CQRS, Data-Oriented Programming, SOLID...

Basically, I tried to break the API into functional endpoints / use cases, so you have in the endpoint path clear indication on what it is doing. We can have specific actions, Commands and Queries, and they will have again very specialised resquest and response payload, with the most meaningfull data for each action.

With that we reduce coupling and allows the API to evolve and introduce different technologies or features with minimum impact in the overal system.

Reply:
Many things you pointed are valid, but they do not cover business scenarios.

Like `POST /application/{id}/members` or `DELETE /application/{id}/members/{memberId}`.

The first endpoint does not reflect the idea that a member is not directly added to the application, but actually a notification should be send to the member and they have to accept the invitation, and only then it's added to the application.

The second does not reflect the idea that a reason must be provided for the member to be removed, and in a RESTfull API DELETE verb usually does not come with a body.

Reply:
Indeed, for special cases like` invite-member` or delete member with body, deviating makes sense. 

However, for endpoints where it's applicable, shouldn't we adhere to the REST convention? 

I personally have preference for the REST convention because it simplifies naming; it's straightforward and intuitive, You don't have to think much. it just comes naturally

Nevertheless, I don't have a strong opinion on this matter. I'm open to using custom routes as well

Reply:
Indeed, I will follow REST convention whenever is possible, for intance, all `GET` endpoints are RESTful. 

And I changed: 
- `POST /applications/{id}/add-attachement` to `POST /applications/{id}/attachments`.
- `POST /applications/{id}/add-event` to `POST /applications/{id}/events`.

All the other endpoints do not fit on RESTful convention because they have extra business steps.

The only special case for me is `/applications/{id}/save-forms-and-duos`. It could be simply `PUT /applications/{id}/forms` and `PUT /applications/{id}/duos`, but I don't know the impact on the integration with REMS. In REMS both forms and DUO codes are persisted together.